source/common/http/conn_manager_impl.cc

```diff
@@ -1396,18 +1396,52 @@ void ConnectionManagerImpl::ActiveStream::snapScopedRouteConfig() {
   }
 }
 
-void ConnectionManagerImpl::ActiveStream::refreshCachedRoute() {
-  Router::RouteConstSharedPtr route;
-  if (request_headers_ != nullptr) {
-    if (connection_manager_.config_.isRoutable() &&
-        connection_manager_.config_.scopedRouteConfigProvider() != nullptr) {
-      // NOTE: re-select scope as well in case the scope key header has been changed by a filter.
-      snapScopedRouteConfig();
+void ConnectionManagerImpl::ActiveStream::refreshCachedRoute() { refreshCachedRoute(nullptr); }
+
+void ConnectionManagerImpl::ActiveStream::refreshCachedRoute(Router::RouteCallbackSharedPtr cb) {
+  if (request_headers_ == nullptr) {
+    cacheRoute(nullptr);
+    if (cb != nullptr) {
```

Comment:
It's very confusing to me that we are calling the callback here and not doing anything with the result. I think if you follow my advice about having both the existing version of route and the new version return the route, the internals should be a common impl where the existing function just passes a null callback.

Reply:
I am confused, I have indeed overloaded the function so existing call indeed pass null callback

Reply:
It's hard for me to follow the logic right now. I think that when you have both functions return the route you will see that they can more easily be implemented in terms of each other. Let's start there and then we can circle back. Thank you!